application/src/main/java/org/thingsboard/server/service/entitiy/otaPackageController/DefaultTbOtaPackageService.java

```diff
@@ -0,0 +1,109 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.thingsboard.server.common.data.EntityType;
+import org.thingsboard.server.common.data.OtaPackage;
+import org.thingsboard.server.common.data.OtaPackageInfo;
+import org.thingsboard.server.common.data.SaveOtaPackageInfoRequest;
+import org.thingsboard.server.common.data.audit.ActionType;
+import org.thingsboard.server.common.data.exception.ThingsboardException;
+import org.thingsboard.server.common.data.id.OtaPackageId;
+import org.thingsboard.server.common.data.id.TenantId;
+import org.thingsboard.server.common.data.ota.ChecksumAlgorithm;
+import org.thingsboard.server.queue.util.TbCoreComponent;
+import org.thingsboard.server.service.entitiy.AbstractTbEntityService;
+import org.thingsboard.server.service.security.model.SecurityUser;
+
+import java.nio.ByteBuffer;
+
+@Service
+@TbCoreComponent
+@AllArgsConstructor
+@Slf4j
+public class DefaultTbOtaPackageService extends AbstractTbEntityService implements TbOtaPackageService {
+    @Override
+    public OtaPackageInfo save(SaveOtaPackageInfoRequest saveOtaPackageInfoRequest, SecurityUser user) throws ThingsboardException {
+        TenantId tenantId = saveOtaPackageInfoRequest.getTenantId();
+        ActionType actionType = saveOtaPackageInfoRequest.getId() == null ? ActionType.ADDED : ActionType.UPDATED;
+        try {
+            OtaPackageInfo savedOtaPackageInfo = otaPackageService.saveOtaPackageInfo(new OtaPackageInfo(saveOtaPackageInfoRequest), saveOtaPackageInfoRequest.isUsesUrl());
+            notificationEntityService.notifyEntity(tenantId, savedOtaPackageInfo.getId(), savedOtaPackageInfo, null,
+                    actionType, user, null);
+            return savedOtaPackageInfo;
+        } catch (Exception e) {
+            notificationEntityService.notifyEntity(tenantId, emptyId(EntityType.OTA_PACKAGE), saveOtaPackageInfoRequest, null,
+                    actionType, user, e);
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public void delete(OtaPackageInfo otaPackageInfo, SecurityUser user) throws ThingsboardException {
+        TenantId tenantId = otaPackageInfo.getTenantId();
+        OtaPackageId otaPackageId = otaPackageInfo.getId();
+        try {
+            otaPackageService.deleteOtaPackage(tenantId, otaPackageId);
+            notificationEntityService.notifyEntity(tenantId, otaPackageId, otaPackageInfo, null,
+                    ActionType.DELETED, user, null, otaPackageInfo.getId().toString());
+        } catch (Exception e) {
+            notificationEntityService.notifyEntity(tenantId, emptyId(EntityType.OTA_PACKAGE), null, null,
+                    ActionType.DELETED, user, e, otaPackageInfo.getId().toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public OtaPackageInfo saveOtaPackageData(OtaPackageInfo otaPackageInfo, String checksum, ChecksumAlgorithm checksumAlgorithm,
```

Comment:
remove the Exception e, you should catch exceptions here like in other entyty services.

Reply:
Ok